### ISSUE TITLE
Add feature: Import/Export Trello/Wekan board to/from Sandstorm Wekan.

### ISSUE THREAD

**Issue by xet7** [wekan/wekan#799]
_From @xet7 on December 16, 2016 11:14_

Original request from @gaconkzk at Wekan Gitter.

_Copied from original issue: wefork/wekan#51_

**Comment by xet7** [comment]
_From @tml on January 18, 2017 16:14_

With the Trello purchase, I suspect this might be something that a number of people become interested in rather quickly.

**Comment by xet7** [comment]
@tml 

In standalone Wefork it's already possible to import Trello export JSON files, including attachment images. For example using VirtualBox appliance at https://wekan.xet7.org

This feature is not enabled for Sandstorm Wefork.

**Comment by xet7** [comment]
_From @ocdtrekkie on January 18, 2017 16:20_

If the import from Trello requires actually logging into Trello, rather than importing like an export file, that'd probably be why, since it requires a network opening to do, and Sandstorm sandboxes the application. Future updates to Sandstorm should make this easier to do securely.

**Comment by xet7** [comment]
Checklists are not imported yet, because that feature #11 is not yet merged to Wefork. Some other features can be missing also.

**Comment by xet7** [comment]
_From @tml on January 18, 2017 16:30_

@ocdtrekkie I have an export file from Trello, and a standalone Wekan knows how to deal with those; there's just no opportunity to provide that information to the Wekan grain created by Sandstorm.

**Comment by xet7** [comment]
_From @ocdtrekkie on January 18, 2017 16:32_

Should definitely tag @dwrensha on this one then.

**Comment by tml** [comment]
This seems to be a duplicate of #800?

**Comment by xet7** [comment]
@tml

Thanks, I closed #800 now. Duplicate came when I used https://github-issue-mover.appspot.com/ to move issues from Wefork to Wekan

**Comment by xet7** [comment]
@FiloSottile said at #549 
I'd love this, too. My understanding from #113 is that #430 should make this possible.

I have not checked those yet.

**Comment by brevilo** [comment]
If it means anything: +1 👍

**Comment by ajrossnz** [comment]
Really looking forward to this. As wekan on its own doesn't have LDAP, we can't use that, so waiting on this feature before we go for implementation here.

**Comment by xet7** [comment]
@ajrossnz @brevilo 

Please do not wait. In Open Source, although current code is available for free, the way to use it in your company is to have a programmer in your company implementing missing features and submitting them as pull requests.

[According to Open Hub](https://www.openhub.net/p/wekan), Wekan code is only about 7000 lines so it's very small when comparing to other software. With git history viewer like gitk it's possible to see how different features are implemented.

To implement this, check current features for exporting and importing, and have the feature enabled in Sandstorm by using environment variable isSandstorm = true like is at wekan/sandstorm.js .

In wiki there is [Developer Documentation](https://github.com/wekan/wekan/wiki/Developer-Documentation).

Thanks!

**Comment by xet7** [comment]
Currently dropdown menu on Wekan looks like this:

![4-menu](https://user-images.githubusercontent.com/15545/28243471-d26d43ce-69d3-11e7-9c4c-5b32ce44049b.png)

In addition to https://github.com/wekan/wekan/pull/1134 for Import to work on Sandstorm @GhassenRjab would need to add:

1) Check that isSandstorm variable is true https://github.com/wekan/wekan/blob/devel/sandstorm.js

2) Add to board menu above Export also Import

3) When Import is clicked, there would be paste screen and warning above it that imported board will delete all existing data on board and replace it with imported board.

I don't know is mapping of imported board users to Sandstorm users somehow different than in standalone Wekan. @dwrensha do you know about this?

Also on same menu where is [Outgoing Webhooks](https://github.com/wekan/wekan/pull/1119). Is there a way to make it work on Sandstorm?

**Comment by xet7** [comment]
@ocdtrekkie @dwrensha @kentonv

On menu abobe there is Archived boards.

Currently on Wekan the All boards link is hidden, and only one board is used on Sandstorm.

When importing new Wekan board, should current board be archived, and new board be made visible?

Then on Board menu / Archived Items there would be Cards, Lists and Boards.

Or should there instead be:
- All boards page and similar functionality like in standalone Wekan?
- Ability to copy cards/lists etc to some other board on the same grain, and also to other grain at Sandstorm?

**Comment by ocdtrekkie** [comment]
The ideal Sandstorm way to handle multiple boards would be to continue with Wekan only containing one board, but that trying to copy cards or lists to another board would bring up a Sandstorm powerbox to select a destination Wekan. That ensures the security isolation continues to work well, that people who can access one board can't tamper with another board even if there's a security issue.

If you can archive the old board when importing a Wekan board, that might be a smart choice just in case someone ignores the warning and overwrites the content accidentally. Presumably someone could dig into the grain backup and pull the lost data out of it with some effort.

**Comment by xet7** [comment]
@ocdtrekkie 

I think in that case, that when importing new Wekan board:
- current board should be archived
- new board be made visible
- on Board menu / Archived Items there would be Cards, Lists and Boards, with options to restore and delete board.

**Comment by xet7** [comment]
Also when restoring archived board, current board should be archived. Only deletion is from archived boards.

**Comment by kentonv** [comment]
FWIW, the ideal way to handle "imports" on Sandstorm would be to have a separate new-grain action. Currently there is only "New board", but you could also have "Import board", which launches directly into an import UI. It feels a bit awkward to create a blank board and then import over it.

However, I'm not sure it's worth your time to try to make this work.

**Comment by xet7** [comment]
Hmm, I don't know yet. There will be additional import export options and they could be more easily found on Wekan Board menu. AFAIK something (maybe restoring board?) was already left out from Sandstorm import export because of complexity to implement it.

**Comment by ocdtrekkie** [comment]
Currently EtherCalc also overwrites the existing content as the import method. It's probably not ideal, but its better than nothing for now.

**Comment by ocdtrekkie** [comment]
As a note, if the packaging hackery got smoothed out so people could build test packages for Sandstorm without David, it'd be easy for me to help people through setting up the separate Sandstorm command and such. But as it stands we have no way to test that.

**Comment by xet7** [comment]
Automatic builds for Sandstorm would be nice, like there is currently for Docker and Ubuntu snap edge.

**Comment by ocdtrekkie** [comment]
In order to do that and it work in production, we'd also need to switch the signing key for Wekan away from David's key (might be prudent anyways since he doesn't seem super interested in building this regularly), and your build server would need to have the new key.

But at the very least, if we had a routine/reproducible way for people to build test packages of Wekan for Sandstorm, it'd be possible to test Sandstorm features on it. (There's more I'd like to do, like enabling access to the API and stuff. Someday hopefully using the Powerbox to move cards and lists between boards, etc.)

If I could package Wekan, I'd be more than happy to work with anyone who needs help implementing Sandstorm features, like a separate "Import board" flow.

**Comment by xet7** [comment]
Code for Export and import boards in Sandstorm is included in today's Wekan release tagged v0.30 .

**Comment by xet7** [comment]
I did make my own Sandstorm key with:

```
spk keygen
```
and add it's public key it displayed after that command to sandstorm-pkgdef.capnp.

Then I generated wekan-0.32.spk on Ubuntu 14.04 64bit on VirtualBox:

```
meteor-spk pack wekan-0.32.spk
```

When I tried to upload it to local Sandstorm, I got this error:

```
Upload failed: 500 Internal Server Error.
Unpacking SPK failed; is it valid?
```

**Comment by ocdtrekkie** [comment]
Did you try `meteor-spk dev` and using the app locally in dev mode? Does that work?

Also, did you do anything with the pgpSignature in there? It signs the app key, and so if you didn't re-sign the app, Sandstorm may think the SPK is invalidly signed. I don't know if that would cause that error or not, mind you.

But whereas packing and unpacking SPKs cares about keys and signing, dev mode does not, so is a good first place to start.

**Comment by xet7** [comment]
I wrote instructions how I did get [newerst Wekan built for Sandstorm](https://github.com/wekan/wekan-maintainer/wiki/Building-Wekan-for-Sandstorm).

@GhassenRjab 

With instructions above, could you test how to Export and Import Wekan board correctly? I tried to setup names for all labels, tried to map users, and got errors that:

- JSON is not formatted correctly
- I did not map users correctly

Maybe user mapping in Sandstorm is different?

**Comment by ocdtrekkie** [comment]
One note: The appVersion must always be higher, but it doesn't matter if they get skipped. It's just an ordinal integer. The app market will reject an appVersion equal to or lower than the published version. That, like the signing/appId stuff won't be of any material importance to people doing testing, since we'll make sure the metadata is all right before we publish.

**Comment by ocdtrekkie** [comment]
I'm curious if @GhassenRjab is around and interested in tackling this issue in the near future, or if we should be trying to package Wekan for Sandstorm with import disabled so that all of the other new fixes and features can be pushed out.

**Comment by ocdtrekkie** [comment]
Alright. Make sure all of the GPG metadata is changed to Wekan Team before publishing, as per David's request. (I would remove the Upstream Author line entirely from the package definition file.)

When you use spk publish, it should check that everything squares up right before putting it in the experimental market at https://apps.sandstorm.io/?experimental=true

We can keep it available there until all of the important issues are all tested, upon which we can release it to everyone else.

**Comment by xet7** [comment]
@GhassenRjab 

Works:
1) Standalone Wekan Export => Standalone Wekan Import
- But: Start date is not imported. Only Due Date.

2) Standalone Wekan Export => Sandstorm Wekan Import
- But: Start date is not imported. Only Due Date.

Does not work:
1) Standstorm Wekan Export => Standalone Wekan Import
- Something wrong in Sandstorm Wekan JSON file

2) Standalone Wekan Export => Standstorm Wekan Import => Standstorm Wekan Export
- JSON file size is zero.

Exported JSON files in .zip below.

[wekan-export.zip](https://github.com/wekan/wekan/files/1227136/wekan-export.zip)

**Comment by xet7** [comment]
@GhassenRjab 

[Info how to test on Sandstorm](https://github.com/wekan/wekan-maintainer/wiki/Building-Wekan-for-Sandstorm)

**Comment by xet7** [comment]
TODO: Test does unreleased version of newest Wekan for Sandstorm has fix for #541 already.

**Comment by xet7** [comment]
I tested importing Trello JSON on Wekan Sandstorm version. It seems to work, but attachment image is not imported, most likely because Wekan grain does not have Sandstorm-compatible outside network access implemented.

@GhassenRjab today tested those JSON files in above wekan-export.zip file. In the first file, one of the activities (addMember) has a userId: null. This will cause the interruption of the import.
He will see if this happens with standalone Wekan.

**Comment by xet7** [comment]
I tested .spk file in Sandstorm. Upgrade of existing Wekan did work. When I then imported board from Trello JSON file, images did not import. From Archives menu I was able to restore previous board that was in Sandstorm before importing board from Trello.

**Comment by xet7** [comment]
About that Copy URL of a card #1188 feature, button in Wekan Sandstorm version copies this kind of URL:

```
/b/sandstorm/libreboard/RANDOMTEXT1
```

While full URL to Sandstorm is:

```
https://something.sandcats.io/grain/RANDOMTEXT2/b/sandstorm/libreboard/RANDOMTEXT1
```

**Comment by ocdtrekkie** [comment]
FWIW, it may be worth summarizing the standing issues for Wekan on Sandstorm and then determining whether or not those are blocking issues. I know in my case, images being imported isn't a big deal, very few of the Trello/Wekan boards I've ever used had any.

Regarding the Copy URL of a card feature, the issue there is that Sandstorm doesn't give the grain's URL to the app.

**Comment by xet7** [comment]
@GhassenRjab and I just got import export working on Wekan stanalone and Sandstorm, I just need to make proper release of Wekan and figure out changing .spk key.

**Comment by ocdtrekkie** [comment]
Any issues following the instructions in the publishing guide, just let me know.

**Comment by AlexanderZobkov** [comment]
Regarding #1188, probably it make sence to create a new issue (telling that #1188 does not work in Sandstorm) or re-open the old one, for the following reason:
URLs like https://something.sandcats.io/grain/RANDOMTEXT2/b/sandstorm/libreboard/RANDOMTEXT1 (in my case http://RANDOMTEXT2hostname.domain/b/sandstorm/libreboard/RANDOMTEXT1) that could be taken by using your brower capabilies  does not work for others. they get something like 
```Error: Unauthorized [403]
    at Proxy.makeContext (server/proxy.js:1660:15)
    at server/proxy.js:1899:19
    at /programs/server/npm/node_modules/meteor/promise/node_modules/meteor-promise/fiber_pool.js:32:39```

**Comment by ocdtrekkie** [comment]
As a note, there's an experimental Sandstorm release of Wekan here: https://apps.sandstorm.io/app/m86q05rdvj14yvn78ghaxynqz7u2svw6rnttptxx49g1785cdv1h?experimental=true

Please run some tests (be safe, don't kill your production server), and let us know if things look good. We know there are some issues still, but it should be by and large a major step up from the last Wekan release on Sandstorm.

**Comment by xet7** [comment]
Wekan v0.34 is now available at official Sandstorm app market. It's not in experimental anymore.
